Library/Homebrew/utils/cp.rb

```diff
@@ -0,0 +1,50 @@
+# typed: true
+# frozen_string_literal: true
+
+require "extend/os/cp"
+require "system_command"
+
+module Utils
+  # Helper functions for interacting with the `cp` command.
+  module Cp
+    class << self
+      sig {
+        params(
+          source:  T.any(String, Pathname, T::Array[T.any(String, Pathname)]),
+          target:  T.any(String, Pathname),
+          sudo:    T::Boolean,
+          verbose: T::Boolean,
+          command: T.class_of(SystemCommand),
+        ).returns(SystemCommand::Result)
+      }
+      def copy_with_attributes(source, target, sudo: false, verbose: false, command: SystemCommand)
+        # On macOS, `cp -p` guarantees to preserve extended attributes (including quarantine
+        # information) in addition to file mode. Other implementations like coreutils does not
+        # necessarily guarantee the same behavior, but that is fine because we don't really need to
+        # preserve extended attributes except when copying Cask artifacts.
+        command.run! "cp", args: ["-p", *extra_flags, *source, target], sudo:, verbose:
+      end
+
+      sig {
+        params(
+          source:  T.any(String, Pathname, T::Array[T.any(String, Pathname)]),
+          target:  T.any(String, Pathname),
+          sudo:    T::Boolean,
+          verbose: T::Boolean,
+          command: T.class_of(SystemCommand),
+        ).returns(SystemCommand::Result)
+      }
+      def copy_recursive_with_attributes(source, target, sudo: false, verbose: false, command: SystemCommand)
```

Comment:
```suggestion
  module Copy
    class << self
      sig {
        params(
          source:  T.any(String, Pathname, T::Array[T.any(String, Pathname)]),
          target:  T.any(String, Pathname),
          sudo:    T::Boolean,
          verbose: T::Boolean,
          command: T.class_of(SystemCommand),
        ).returns(SystemCommand::Result)
      }
      def with_attributes(source, target, sudo: false, verbose: false, command: SystemCommand)
        # On macOS, `cp -p` guarantees to preserve extended attributes (including quarantine
        # information) in addition to file mode. Other implementations like coreutils does not
        # necessarily guarantee the same behavior, but that is fine because we don't really need to
        # preserve extended attributes except when copying Cask artifacts.
        command.run! "cp", args: ["-p", *extra_flags, *source, target], sudo:, verbose:
      end

      sig {
        params(
          source:  T.any(String, Pathname, T::Array[T.any(String, Pathname)]),
          target:  T.any(String, Pathname),
          sudo:    T::Boolean,
          verbose: T::Boolean,
          command: T.class_of(SystemCommand),
        ).returns(SystemCommand::Result)
      }
      def recursive_with_attributes(source, target, sudo: false, verbose: false, command: SystemCommand)
```

thoughts on this naming?

Reply:
Just for the record, the new method names are tailored to fully-qualified method calls like `Utils::Cp.with_attributes` and I think we'd need to add aliases if we are to use the module as a mixin in the future.